text/2540-formal-function-parameter-attributes.md

```diff
@@ -0,0 +1,242 @@
+- Feature Name: formal_function_param_attrs
+- Start Date: 2018-10-14
+- RFC PR: 
+- Rust Issue: 
+
+# Summary
+[summary]: #summary
+
+This RFC proposes to allow attributes in formal function parameter position.
+
+# Motivation
+[motivation]: #motivation
+
+Having attributes on formal function parameters allows for certain different use cases.
+
+## Example: Handling of unused parameter
+
+In today's Rust it is possible to prefix the name of an identifier to silence the compiler about it being unused.
+With attributes in formal function parameter position we could have an attribute like `#[unused]` that explicitely states this for a given parameter.
+
+```rust
+fn foo(#[unused] bar: u32) -> bool;
+```
+
+Instead of
+
+```rust
+fn foo(_bar: u32) -> bool
+```
+
+This would better reflect the explicit nature of Rust compared to the underscore prefix as of today.
+
+## Example: Low-level code
+
+For raw pointers that are oftentimes used when operating with C code one could provide the compiler with additional information about the set of parameters.
```

Comment:
Still hypothetical unless actually proposed.

Reply:
Should I change ideas around use cases for this proposal so that the ideas are directly implementable and usable? If so I (we) have to think about other examples.

Reply:
Thinking about more use cases is always a good thing to strengthen the motivation.

As for the text you've written that is not about proc macros, I'd mostly do an accounting maneuver, move them to the section on procedural macros where such a macro could do this and then reword to fit that section.

Attributes can also be used by static analysis tools, so they don't have to have computational meaning directly in code; tools can give them meaning.

Reply:
Good point about the tooling.
Also good idea to reformulate a few examples for proc macros.